src/set_theory/ordinal/notation.lean

```diff
@@ -791,6 +791,142 @@ begin
       exact (repr_opow_aux₂ _ ad a00 al _ _).2 } }
 end
 
+/-- Given an ordinal, returns `inl none` for `0`, `inl (some a)` for `a+1`, and
+  `inr f` for a limit ordinal `a`, where `f i` is a sequence converging to `a`. -/
+def fundamental_seq : onote → option onote ⊕ (ℕ → onote)
+| zero := sum.inl none
+| (oadd a m b) :=
+  match fundamental_seq b with
+  | sum.inr f := sum.inr (λ i, oadd a m (f i))
+  | sum.inl (some b') := sum.inl (some (oadd a m b'))
+  | sum.inl none := match fundamental_seq a, m.nat_pred with
+    | sum.inl none, 0 := sum.inl (some zero)
+    | sum.inl none, m+1 := sum.inl (some (oadd zero m.succ_pnat zero))
+    | sum.inl (some a'), 0 := sum.inr (λ i, oadd a' i.succ_pnat zero)
+    | sum.inl (some a'), m+1 := sum.inr (λ i, oadd a m.succ_pnat (oadd a' i.succ_pnat zero))
+    | sum.inr f, 0 := sum.inr (λ i, oadd (f i) 1 zero)
+    | sum.inr f, m+1 := sum.inr (λ i, oadd a m.succ_pnat (oadd (f i) 1 zero))
+    end
+  end
+
+theorem exists_lt_add {α} [hα : nonempty α] {o : ordinal} {f : α → ordinal}
+  (H : ∀ a, a < o → ∃ i, a < f i) (b : ordinal) (a) (h : a < b + o) : ∃ i, a < b + f i :=
+begin
+  cases lt_or_le a b with h h',
+  { obtain ⟨i⟩ := id hα, exact ⟨i, h.trans_le (le_add_right _ _)⟩ },
+  { rw [← ordinal.add_sub_cancel_of_le h', add_lt_add_iff_left] at h,
+    refine (H _ h).imp (λ i H, _),
+    rwa [← ordinal.add_sub_cancel_of_le h', add_lt_add_iff_left] }
+end
+
+theorem exists_lt_mul_omega' {o : ordinal} (a) (h : a < o * ω) : ∃ i : ℕ, a < o * ↑i + o :=
+begin
+  obtain ⟨i, hi, h'⟩ := (lt_mul_of_limit omega_is_limit).1 h,
+  obtain ⟨i, rfl⟩ := lt_omega.1 hi,
+  exact ⟨i, h'.trans_le (le_add_right _ _)⟩
+end
+
+local infixr ^ := @pow ordinal ordinal ordinal.has_pow
+theorem exists_lt_omega_opow' {α} {o b : ordinal}
+  (hb : 1 < b) (ho : o.is_limit) {f : α → ordinal}
+  (H : ∀ a, a < o → ∃ i, a < f i) (a) (h : a < b ^ o) : ∃ i, a < b ^ f i :=
+begin
+  obtain ⟨d, hd, h'⟩ := (lt_opow_of_limit (ordinal.zero_lt_one.trans hb).ne' ho).1 h,
+  exact (H _ hd).imp (λ i hi, h'.trans $ (opow_lt_opow_iff_right hb).2 hi)
+end
+
+/-- The property satisfied by `fundamental_seq o`:
+  * `inl none` means `o = 0`
+  * `inl (some a)` means `o = a.succ`
+  * `inr f` means `o` is a limit ordinal and `f` is a
+    strictly increasing sequence which converges to `o` -/
+def fundamental_seq_prop (o : onote) : option onote ⊕ (ℕ → onote) → Prop
+| (sum.inl none) := o = 0
+| (sum.inl (some a)) := o.repr = a.repr.succ ∧ (o.NF → a.NF)
+| (sum.inr f) := o.repr.is_limit ∧
+  (∀ i, f i < f (i + 1) ∧ f i < o ∧ (o.NF → (f i).NF)) ∧
```

Comment:
Why not `strict_mono f ∧ sup.{0 0} (repr ∘ f) = o.repr`?

Reply:
because I don't want to do the big induction proof multiple times so everything of interest about the fundamental sequence is packed in this definition. It is written to optimize the induction proof, you can deduce stuff like that as a consequence.